accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,969 @@
+
+An inline class declaration _DV_ named `Inline` may declare one or
+more constructors. A constructor which is declared in an inline class
+declaration is also known as an _inline class constructor_.
+
+*The purpose of having an inline class constructor is that it bundles
+an approach for building an instance of the representation type of an
+inline declaration _DV_ with _DV_ itself, which makes it easy to
+recognize that this is a way to obtain a value of that inline type. It
+can also be used to verify that an existing object (provided as an
+actual argument to the constructor) satisfies the requirements for
+having that inline type.*
+
+A compile-time error occurs if an inline class constructor includes a
+superinitializer. *That is, a term of the form `super(...)` or
+`super.id(...)` as the last element of the initializer list.*
+
+*In the body of a generative inline class constructor, the static type
+of `this` is the same as it is in any instance member of the inline,
+that is, `Inline<X1 .. Xk>`, where `X1 .. Xk` are the type parameters
+declared by `Inline`.*
+
+An instance creation expression of the form
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;(...)</code>
+or
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;.name(...)</code>
+is used to invoke these constructors, and the type of such an expression is
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>.
+
+*In short, inline class constructors appear to be very similar to
+constructors in regular classes, and they correspond to the situation
+where the enclosing class has a single, non-late, final instance
+variable, which is initialized according to the normal rules for
+constructors (in particular, it must occur by means of `this.id` or in
+an initializer list).*
+
+
+### Composing Inline Classes
+
+This section describes the effect of including a clause derived from
+`<interfaces>` in an inline class declaration. We use the phrase
+_the implements clause_ to refer to this clause.
+
+*The rationale is that the set of members and member implementations
+of a given inline class may need to overlap with that of other inline
+classes. The implements clause allows for implementation reuse by
+putting some shared members in an inline class `V`, and including `V`
+in the implements clause of several inline class declarations
+<code>V<sub>1</sub> .. V<sub>k</sub></code>, thus "inheriting" the
+members of `V` into all of <code>V<sub>1</sub> .. V<sub>k</sub></code>
+without code duplication.*
+
+*The reason why this mechanism uses the keyword `implements` rather
+than `extends` to declare a relation that involves "inheritance" is
+that it has a similar semantics as that of extension members (in that
+they are statically resolved).*
+
+Assume that _DV_ is an inline class declaration named `Inline`, and
+`V1` occurs as one of the `<type>`s in the `<interfaces>` of _DV_. In
+this case we say that `V1` is a _superinterface_ of _DV_.
```

Comment:
I don't think I see specified: what is the behavior when no `implements` clause is supplied? Is `Object` the singular superinterface?

I'm also asking whether an inline class without an `implements` clause has members like `runtimeType` and `operator ==` and `toString`.

Reply:
This has been made explicit in the updated explanation about inline method invocations. The members of `Object` can be invoked on any expression whose static type is an inline class type.

(The intention is that this can be generalized to allow an inline class to provide access to any superinterface of the representation type. I'm keeping 'superinterfaces' of the inline class separate because they are all inline classes, and we do not wish to imply that they carry constraints like 'correct override'. So the relationship to `Object` is different than the relationship between two inline classes.)

I added an extra paragraph here specifying that the set of superinterfaces is empty when there is no `implements` clause.

The result is that an expression whose static type is an inline class _does_ admit invocations of `toString()` etc.

Reply:
News: Leaf and I discussed the treatment of `Object` interface names (`toString` etc.) and agreed that it's a safer bet to make them errors for now: With this change, an inline class can _not_ declare a member named `toString`.

The point is that it is very unlikely that an inline member shadowing an `Object` member will work as intended. If operator `==` is shadowed then sets and maps will still call the instance member; shadowing `noSuchMethod` is completely useless because basically (or strictly?) every single, meaningful invocation of `noSuchMethod` is done by code that doesn't know the static type of the receiver; and so on. Even with `toString`, where a custom version in an inline class might be useful, it's not likely to work as intended. Just consider the fact that _every_ string interpolation would have to use `'${x.toString()}'` rather than `'$x'`, and we would have to ensure that the static type of `x` is actually the given inline class in order to get the inline method when calling `x.toString()`.